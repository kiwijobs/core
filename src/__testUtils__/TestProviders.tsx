import React, { FC, Suspense } from 'react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../theme';

export const TestProviders: FC<{}> = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Suspense fallback="mocked loading...">{children}</Suspense>
    </ThemeProvider>
  );
};

export const wrapper = TestProviders; // React Testing Library friendly name.
