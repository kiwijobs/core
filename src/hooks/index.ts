export { useBodyLock } from './useBodyLock';
export * from './useTheme';
