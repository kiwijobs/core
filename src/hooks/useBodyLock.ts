import { useEffect, useRef } from 'react';

// prevent from enabling body scroll i.e. modal in modal
// enable only when all components with body lock are unmounted
let count = 0;

const disableBodyScroll = () => {
  const { scrollY } = window;
  const top = `-${scrollY}px`;

  document.querySelector('html')?.setAttribute('data-scroll-lock', '1');
  document.body.style.top = top;
  count++;
};

const enableBodyScroll = () => {
  count--;
  if (count > 0) {
    return;
  }
  const scrollY = document.body.style.top;
  document.querySelector('html')?.removeAttribute('data-scroll-lock');
  document.body.style.top = '';
  window.scrollTo(0, parseInt(scrollY || '0') * -1);
};

export const useBodyLock = () => {
  const ref = useRef(null);

  useEffect(() => {
    const current = ref.current;

    if (!current) {
      return;
    }

    disableBodyScroll();

    return () => enableBodyScroll();
  }, [ref]);

  return [ref];
};
