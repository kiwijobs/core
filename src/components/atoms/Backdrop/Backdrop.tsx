import React from 'react';
import { Box, BoxProps } from '../../quarks';
import { rgba } from '../../../theme/colors';
import { useBodyLock } from '../../../hooks';

export const Backdrop = ({ sx, ...props }: BoxProps) => {
  const [ref] = useBodyLock();

  return (
    <Box
      ref={ref}
      sx={{
        position: 'fixed',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        bg: rgba(0.4).dark,
        zIndex: -1,
        ...sx,
      }}
      {...props}
    />
  );
};
