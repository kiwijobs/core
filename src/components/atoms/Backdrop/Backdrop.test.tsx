import React from 'react';
import { wrapper } from '../../../__testUtils__';
import { Backdrop } from './Backdrop';
import { render } from '@testing-library/react';
import * as hooks from '../../../hooks';

Object.defineProperty(window, 'scrollTo', { value: jest.fn(), writable: true });

describe('Backdrop', () => {
  it('Should match snapshot', () => {
    expect(render(<Backdrop />, { wrapper }).container).toMatchSnapshot();
  });

  it('Should use body lock', () => {
    const mockBodyLock = jest.fn(() => []);
    jest.spyOn(hooks, 'useBodyLock').mockImplementation(mockBodyLock);

    render(<Backdrop />, { wrapper });

    expect(mockBodyLock).toHaveBeenCalledWith();
    expect(mockBodyLock).toHaveBeenCalledTimes(1);
  });
});
