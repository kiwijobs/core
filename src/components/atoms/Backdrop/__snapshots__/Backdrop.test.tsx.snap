// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop Should match snapshot 1`] = `
.c0 {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(40,50,56,0.4);
}

<div>
  <div
    class="c0"
  />
</div>
`;
